server/src/main/java/io/druid/server/namespace/NamespacedExtractionModule.java

```diff
@@ -0,0 +1,448 @@
+
+    binder
+        .bind(ExtractionNamespaceFunctionFactory.class)
+        .annotatedWith(Names.named(JDBCExtractionNamespace.class.getCanonicalName()))
+        .to(JDBCExtractionNamespaceFunctionFactory.class)
+        .in(LazySingleton.class);
+    binder
+        .bind(ExtractionNamespaceFunctionFactory.class)
+        .annotatedWith(Names.named(URIExtractionNamespace.class.getCanonicalName()))
+        .to(URIExtractionNamespaceFunctionFactory.class)
+        .in(LazySingleton.class);
+
+    binder
+        .bind(Key.get(Object.class, Names.named(NAMESPACE_ENABLED_KEY)))
+        .toInstance(new Object());
+
+    LifecycleModule.register(binder, NamespacedKeeper.class);
+  }
+
+  public static boolean isNamespaceEnabled(Injector injector)
+  {
+    return null != injector.getExistingBinding(Key.get(Object.class, Names.named(NAMESPACE_ENABLED_KEY)));
+  }
+
+  @Provides
+  @Named("namespaceExtractionFunctionCache")
+  public ConcurrentMap<String, Function<String, String>> getFnCache()
+  {
+    return fnCache;
+  }
+
+  @Provides
+  @Named("dimExtractionNamespace")
+  @LazySingleton
+  public Function<String, Function<String, String>> getFunctionMaker(
+      @Named("namespaceExtractionFunctionCache")
+      final ConcurrentMap<String, Function<String, String>> fnCache
+  )
+  {
+    return new Function<String, Function<String, String>>()
+    {
+      @Nullable
+      @Override
+      public Function<String, String> apply(final String namespace)
+      {
+        Function<String, String> fn = fnCache.get(namespace);
+        if (fn == null) {
+          throw new IAE("Namespace [%s] not found", namespace);
+        }
+        return fn;
+      }
+    };
+  }
+
+  @ManageLifecycle
+  public static class NamespacedKeeper
+  {
+    private final CuratorFramework curator;
+    private final ZkPathsConfig zkPathsConfig;
+    private final ListeningExecutorService listeningExecutorService = MoreExecutors.listeningDecorator(
+        Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("NamespaceKeeper-%d")
+                .setPriority(Thread.MIN_PRIORITY)
+                .build()
+        )
+    );
+    private final Injector injector;
+    private final NamespaceExtractionCacheManager namespaceExtractionCacheManager;
+    private final ObjectMapper jsonMapper;
+    private final PathChildrenCache pathChildrenCache;
+    private PathChildrenCacheListener listener;
+
+    @Inject
+    public NamespacedKeeper(
+        CuratorFramework curator,
+        ZkPathsConfig zkPathsConfig,
+        @Json ObjectMapper jsonMapper,
+        Injector injector,
+        NamespaceExtractionCacheManager namespaceExtractionCacheManager
+    )
+    {
+      this.curator = curator;
+      this.zkPathsConfig = zkPathsConfig;
+      this.jsonMapper = jsonMapper;
+      this.injector = injector;
```

Comment:
realized I couldn't remove it. The issue is with wiring up stuff in Processing to stuff in Server, and this was the cleanest way I could get it at the time. I suppose I *could* do a multi-map or similar, but that doesn't seem too different than this method except it is more explicit.